AWS CloudWatch Event Sink Implementation

adnanhemani opened the pull request:
<!--
    Possible security vulnerabilities: STOP here and contact security@apache.org instead!

    Please update the title of the PR with a meaningful message - do not leave it "empty" or "generated"
    Please update this summary field:

    The summary should cover these topics, if applicable:
    * the motivation for the change
    * a description of the status quo, for example the current behavior
    * the desired behavior
    * etc

    PR checklist:
    - Do a self-review of your code before opening a pull request
    - Make sure that there's good test coverage for the changes included in this PR
    - Run tests locally before pushing a PR (./gradlew check)
    - Code should have comments where applicable. Particularly hard-to-understand
      areas deserve good in-line documentation.
    - Include changes and enhancements to the documentation (in site/content/in-dev/unreleased)
    - For Work In Progress Pull Requests, please use the Draft PR feature.

    Make sure to add the information BELOW this comment.
    Everything in this comment will NOT be added to the PR description.
-->

As per discussion at this morning's Polaris Community Sync, I am introducing this event listener that will allow users to sink events to AWS CloudWatch Logs - where they can search through events for auditing purposes, if they wish. I do intend to introduce similar changes for Azure and GCP once this is merged so that we have parity across all CSPs - but this will help nail down the exact pattern to make this work.

Below is a screenshot of a sample event sent to CloudWatch:

![Screenshot 2025-06-26 at 7 20 31 PM](https://github.com/user-attachments/assets/25482b53-3117-466b-a168-e27a4f521578)

snazy commented:
Didn't look into the whole PR, but two things I noticed:
* The PR's doing more than mentioned in the summary & description
* There's no (integration) testing for the feature (IIRC there are solutions to test cloudwatch locally)

adnanhemani commented:
I've added a test using LocalStack, which should help emulate CloudWatch. This PR is ready for a full review.

eric-maynard commented on runtime/defaults/src/main/resources/application.properties:
```diff
@@ -126,6 +126,10 @@ polaris.secrets-manager.type=in-memory
 polaris.file-io.type=default
 
 polaris.event-listener.type=no-op
+# polaris.event-listener.type=aws-cloudwatch
+# polaris.event-listener.aws-cloudwatch.log-group=test-group
+# polaris.event-listener.aws-cloudwatch.log-stream=test-stream
+# polaris.event-listener.aws-cloudwatch.region=us-west-2
```
The intent behind these comments is good, but since it's a bit cloud-specific we might be better off leaving it out

adnanhemani replied:
Sure, that makes sense. Where do you recommend we document this instead?

eric-maynard replied:
Maybe if we had an application.properties specific to AWS, such as in the getting started section?

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/EventListenerConfiguration.java:
```diff
@@ -0,0 +1,30 @@
+import java.util.Optional;
+
+public interface EventListenerConfiguration {
+  Optional<String> awsCloudwatchlogGroup();
+
+  Optional<String> awsCloudwatchlogStream();
+
+  Optional<String> awsCloudwatchRegion();
```
How does this work if I want to define my own EventListener -- do I essentially need to fork and modify this class?

Won't this also grow quite large as more listeners get implemented?

adnanhemani replied:
> How does this work if I want to define my own EventListener -- do I essentially need to fork and modify this class?

That's one way to do it. The other would be to create a new interface and have `QuarkusPolarisEventListenerConfiguration` extend the new interface as well. Maybe there's a different way to do it, but I haven't found one yet.

> Won't this also grow quite large as more listeners get implemented?

It could. If we start getting a lot of new listeners, we can always split into multiple interfaces. I'm not doing this yet (it might be premature optimization) but we could very easily do this in the future.

eric-maynard replied:
If possible, I think it would make sense to create a different configuration for each listener -- ideally, in such a way to make it easy for users to add their own listener. This should also help isolate the code of various listeners.

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,228 @@
+      EventListenerConfiguration config, ExecutorService executorService) {
+    this.executorService = executorService;
+
+    this.logStream = config.awsCloudwatchlogStream().orElse("polaris-cloudwatch-default-stream");
+    this.logGroup = config.awsCloudwatchlogGroup().orElse("polaris-cloudwatch-default-group");
+    this.region = Region.of(config.awsCloudwatchRegion().orElse("us-east-1"));
```
Let's make constants out of these

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,228 @@
+
+  @VisibleForTesting
+  public void drainQueue() {
+    List<EventAndTimestamp> drainedEvents = new ArrayList<>();
+    List<InputLogEvent> transformedEvents = new ArrayList<>();
```
If we stick with this approach, we should re-use these objects

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,228 @@
+      drainedEvents.forEach(event -> transformedEvents.add(createLogEvent(event)));
+
+      sendToCloudWatch(transformedEvents);
+    } catch (Exception e) {
```
this catch is too broad imo

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,228 @@
+      sendToCloudWatch(transformedEvents);
+    } catch (Exception e) {
+      LOGGER.error("Error writing logs to CloudWatch: {}", e.getMessage());
+      LOGGER.error("Events not logged: {}", transformedEvents);
```
This could be quite verbose, can we just print one event or some kind of summary?

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,228 @@
+  private void ensureLogGroupAndStream() {
+    try {
+      client.createLogGroup(CreateLogGroupRequest.builder().logGroupName(logGroup).build());
+    } catch (ResourceAlreadyExistsException ignored) {
+    }
```
Might as well add a debug log here

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,228 @@
+      } catch (InvalidSequenceTokenException e) {
+        sequenceToken = getSequenceToken();
+        requestBuilder.sequenceToken(sequenceToken);
+        PutLogEventsResponse retryResponse = client.putLogEvents(requestBuilder.build());
+        sequenceToken = retryResponse.nextSequenceToken();
```
nit: some repeated code here could be refactored out

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,228 @@
+    }
+  }
+
+  private record EventAndTimestamp(String event, long timestamp) {}
```
nit: `EventWithTimestamp`?

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,228 @@
+  @Override
+  public void onAfterCatalogCreated(AfterCatalogCreatedEvent event, CallContext callContext) {
+    try {
+      Map<String, Object> json = objectMapper.convertValue(event.catalog(), Map.class);
```
IIUC this will result in a lot of duplicated code for each event type, which we should try to refactor away. Also, it may make sense to separate these from the actual "logic" of the listener.

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/PolarisEventListener.java:
```diff
@@ -55,4 +57,7 @@ public void onBeforeTaskAttempted(BeforeTaskAttemptedEvent event) {}
   public void onAfterTaskAttempted(AfterTaskAttemptedEvent event) {}
+
+  /** {@link AfterCatalogCreatedEvent} */
+  public void onAfterCatalogCreated(AfterCatalogCreatedEvent event, CallContext callContext) {}
```
This isn't strictly required for the sink implementation, can we consider separating it out? IIUC there is another PR that actually subsumes this change and will make it unnecessary?

adnanhemani replied:
This statement, while true, on its own is not the full picture. In creating this event, you can see the `CallContext` that was added to the function signature - we require that `CallContext` for the rest of the PR.

So, if I remove this event, we no longer have unit tests that work - unless I modify a different event. In which case, I don't really see the opposition to adding this event anymore if we're okay with modifying other events as part of this PR (which will also see changes in other PRs).

If this is a blocker, I can remove this new event - but then this PR will get even bigger with changes to all other events that currently exist. Let me know if there is a strong preference - but I'd prefer to keep this PR as-is and let followup PRs make whatever adjustments are required for all events.

eric-maynard replied:
Got it, that makes sense -- we need working unit tests. My preference would still be to make the event changes separately and then rebase this PR on top of main once those changes merge, but I won't go so far as to say this change must be removed from the current PR since that would break tests.

eric-maynard commented on service/common/src/test/java/org/apache/polaris/service/events/AwsCloudWatchEventListenerTest.java:
```diff
@@ -0,0 +1,264 @@
+      LoggerFactory.getLogger(AwsCloudWatchEventListenerTest.class);
+
+  private static final DockerImageName LOCALSTACK_IMAGE =
+      DockerImageName.parse("localstack/localstack:3.4");
```
I don't know much about localstack but this looks a bit scary, is there a way we can use our existing regtests/docker setup for this?

eric-maynard replied:
If possible, I think we should have a "pure" unit test that mocks whatever needs to be mocked. Having a localstack test is good too, but that does feel more like a regression test.

eric-maynard commented on runtime/service/src/main/java/org/apache/polaris/service/quarkus/events/QuarkusPolarisEventListenerConfiguration.java:
```diff
@@ -20,13 +20,29 @@
+
+  @WithName("aws-cloudwatch.log-group")
+  @Override
+  Optional<String> awsCloudwatchlogGroup();
```
My same comment about extensibility applies; I would expect a new `...Configuration` type per `PolarisEventListener` implementation

eric-maynard replied:
This is probably my biggest remaining concern after the recent changes... I just imagine that with the 3+ event listeners you're proposing these config types will get quite messy.

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+
+  ExecutorService executorService;
+
+  private Future<?> backgroundTask;
```
Maybe we can add an _option_ to immediately emit events? Something like `polaris.event-listener.aws-cloudwatch.async=false` could do it.

re: The discussion on using Quarkus to manage the async tasks I don't have an opinion on that but I imagine others might.

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+        queue.addAll(drainedEvents);
+      }
+    } catch (DataAlreadyAcceptedException e) {
+      LOGGER.debug("Data already accepted: {}", e.getMessage());
```
In this case, isn't it possible only _some_ of the events were already accepted?

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+
+  private volatile boolean running = true;
+
+  ExecutorService executorService;
```
Not thread safe

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+
+  ExecutorService executorService;
+
+  private Future<?> backgroundTask;
```
Not thread safe

eric-maynard replied:
What about it is not thread safe? It's only ever assigned in the PostConstruct; it looks like it could be `final` though

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+  @PreDestroy
+  void shutdown() {
+    running = false;
+    if (backgroundTask != null) {
+      try {
+        backgroundTask.get(10, TimeUnit.SECONDS);
+      } catch (Exception e) {
+        LOGGER.error("Error waiting for background logging task to finish: {}", e.getMessage());
+      }
+    }
```
Relying on timing does not look not safe to me.

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+  }
+
+  // Event overrides below
+  private void queueEvent(PolarisEvent event, CallContext callContext) {
```
This can easily serialize many (100s of) MB. Can AWS CloudWatch handle such big payloads?

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/EventListenerConfiguration.java:
```diff
@@ -0,0 +1,30 @@
+
+import java.util.Optional;
+
+public interface EventListenerConfiguration {
```
How would users provide (different) credentials?

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+
+@ApplicationScoped
+@Identifier("aws-cloudwatch")
+public class AwsCloudWatchEventListener extends PolarisEventListener {
```
This is a _specific_ implementation.
Even if it is rather just a few classes, it deserves at least its own package. Putting this in a separate module is even better, because that prevents CI from running the expensive integration tests for unrelated changes.

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+  private static final String DEFAULT_LOG_GROUP_NAME = "polaris-cloudwatch-default-group";
+  private static final String DEFAULT_REGION = "us-east-1";
+
+  private final BlockingQueue<EventWithTimestamp> queue = new LinkedBlockingQueue<>();
```
This queue is unbounded. It's likely that Polaris will OOM in case of communication issues w/ CloudWatch.

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+  }
+
+  private void processQueue() {
+    List<EventWithTimestamp> drainedEvents = new ArrayList<>();
+    List<InputLogEvent> transformedEvents = new ArrayList<>();
```
The two collections look to be an unnecessary complication and should be simplified.

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+        return;
+      }
+
+      drainedEvents.forEach(event -> transformedEvents.add(createLogEvent(event)));
```
Why is this `createLogEvent()` done here?

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+      drainedEvents.forEach(event -> transformedEvents.add(createLogEvent(event)));
+
+      sendToCloudWatch(transformedEvents);
+    } catch (InterruptedException e) {
```
The purpose of an IE is to _stop_, not to continue.

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+      }
+      LOGGER.error("Number of dropped events: {}", transformedEvents.size());
+      LOGGER.debug("Events not logged: {}", transformedEvents);
+      queue.addAll(drainedEvents);
```
This (and the one above) lead to OOMs. Auth issues and availability impacts will lead to this.

snazy commented on service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,264 @@
+        PutLogEventsRequest.builder()
+            .logGroupName(logGroup)
+            .logStreamName(logStream)
+            .logEvents(events);
```
I would bet that AWS has a limit on the number of events.

snazy commented on service/common/src/test/java/org/apache/polaris/service/events/AwsCloudWatchEventListenerTest.java:
```diff
@@ -0,0 +1,278 @@
+      LoggerFactory.getLogger(AwsCloudWatchEventListenerTest.class);
+
+  private static final DockerImageName LOCALSTACK_IMAGE =
+      DockerImageName.parse("localstack/localstack:3.4");
```
Hard coding dependency versions in source isn't good practice. This should be replaced with the container-spec-helper.

snazy reviewed (changes requested):
There are some thread-safety/concurrency issues and memory leaks in this change, which have to be addressed.

The implementation can also be simplified.

I'd also recommend to move this to its own module to streamline CI and testing in general.

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/aws/cloudwatch/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,275 @@
+
+  // Event overrides below
+  @Override
+  public void onAfterTableRefreshed(AfterTableRefreshedEvent event) {
```
I still think we need to refactor these out so that the listener itself with all its Future-wrangling logic does not become exceedingly long

eric-maynard commented on service/common/src/main/java/org/apache/polaris/service/events/aws/cloudwatch/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,275 @@
+      return;
+    }
+    InputLogEvent inputLogEvent = createLogEvent(eventAsJson, getCurrentTimestamp(callContext));
+    Future<?> future = executorService.submit(() -> sendAndHandleCloudWatchCall(inputLogEvent));
```
Two thoughts:
1. For this listener, I think we should offer a simple synchronous submission mode
2. As we add more listeners, seems like some of this asynchronous handling logic will be duplicated so it might be worth refactoring it out. This can wait until we add more listeners though.

eric-maynard reviewed (commented):
I think this is getting really close. Left a couple of comments on the listener structure.

adnanhemani commented on service/common/src/main/java/org/apache/polaris/service/events/jsonEventListener/JsonEventListener.java:
```diff
@@ -0,0 +1,38 @@
+
+import java.util.HashMap;
+
+public abstract class JsonEventListener extends PolarisEventListener {
```
The map will transform into a JSON string when it sinks, which is why I made it this name. But open to change if you think `MapBasedEventListener` is better?

adnanhemani commented on service/common/src/test/resources/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/Dockerfile-localstack-version:
```diff
@@ -0,0 +1,23 @@
+# Dockerfile to provide the image name and tag to a test.
+# Version is managed by Renovate - do not edit.
+FROM localstack/localstack:3.4
+
```
Removed!

adnanhemani commented on runtime/service/src/main/java/org/apache/polaris/service/quarkus/events/jsonEventListener/aws/cloudwatch/QuarkusAwsCloudWatchConfiguration.java:
```diff
@@ -0,0 +1,53 @@
+@StaticInitSafe
+@ConfigMapping(prefix = "polaris.event-listener.aws-cloudwatch")
+@ApplicationScoped
+public interface QuarkusAwsCloudWatchConfiguration extends AwsCloudWatchConfiguration {
```
Added in next revision!

RussellSpitzer commented on service/common/src/main/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,272 @@
+    for (Future<?> future : futures.keySet()) {
+      if (future.isDone()) {
+        Future.State currFutureState = future.state();
+        EventWithRetry currValue = futures.remove(future);
```
It's also probably safer here to use the view iterators's own "remove" function rather than calling remove on the underlying collection.

sfc-gh-ahemani replied:
The iterator's `remove` function is practically the same as calling `remove` on the map itself: 

```
public boolean remove(Object o) { return map.remove(o) != null; }
```

If we want to call remove on the iterator itself, we will have to restructure the code a bit in a way that I think will make it look even less safe. I'd prefer to keep the `remove` as-is unless there's a concrete reason otherwise. (My quick research makes it seem like there's really not much of a difference).

Let me fix the potential NPE with just making sure that the returned value is not null (i.e. only continue processing this key if the current thread was the first to retrieve+remove the future from the map).

RussellSpitzer commented:
Hi y'all, I'm a bit late to this discussion but i'm a little confused around a few issues

Mostly this centers around the Futures tracking we are doing,

1. Why do we need to do our own app level retries of AWS requests? Doesn't the Cloudwatch SDK provide retries and policy that we can configure if we wanted additional retries? I'm not sure the complexity here is worth while since we don't have any app logic that needs to get redone when a retry happens. I'm not opposed to doing the retries on the app side, i'm just not sure why they are necessary.

2. Why do we need to reap futures at all? Do we not trust Quarkus to shutdown the executor service appropriately? It seems a bit counter-intuitive to me that we would be provided an ExecutorService by the runtime but then in case of emergency we need to manually manage it.

adnanhemani commented:
Thanks for this, @RussellSpitzer. 

1. I added the additional retries for redundancy "best-effort". But I'm also aligned if we don't do this manually and leave it to the AWK SDK.
2. I'm not opposed to not tracking the futures that we added to the ExecutorService - but I had been previously chided for not handling these in other PRs in this repo (not sure I remember by who). @snazy - can you please add your thoughts on this? I'm glad to remove the `futures` ConcurrentHashMap if we do not need to track the futures.

eric-maynard commented:
I think the future-reaping predates the use of the Quarkus ExecutorService -- it can probably go, but it's also not a bad idea to keep a handle on the future's we've submitted.

snazy reviewed (changes requested):
I do not think that the implementation is correct. The concurrency aspects in the code are racy and lead to wrong behavior. That concurrent behavior lacks explicit test coverage, which only provides a _single threaded_ executor, which is not what a production system would use.

Overall, this one implements something that's already present out of the box in the AWS SDK.

adnanhemani commented on runtime/service/src/main/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,276 @@
+    InputLogEvent inputLogEvent = createLogEvent(eventAsJson, getCurrentTimestamp(callContext));
+    if (!synchronousMode) {
+      Future<?> future = executorService.submit(() -> sendAndHandleCloudWatchCall(inputLogEvent));
+      futures.put(future, new EventWithRetry(inputLogEvent, 0));
```
Switched to it.

adnanhemani commented on runtime/service/src/main/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,276 @@
+          }
+        }
+      }
+    }
```
I don't see the race condition, to be honest. But I have changed over to the async CloudWatchClient regardless - PTAL.

adnanhemani commented on runtime/service/src/main/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,276 @@
+        sequenceToken = getSequenceToken();
+        requestBuilder.sequenceToken(sequenceToken);
+        executePutLogEvents(requestBuilder);
+      }
```
This is a newer thing that has come out that I wasn't aware of - thanks! I've made the change to remove the sequence token.

adnanhemani reviewed (commented):
@snazy - I've changed over to using the CloudWatchAsyncClient as you suggested and thus am relying on the AWS SDK as much as possible to work out of the box, including removing manual futures handling and running on the ExecutorService provided to us by the server (AWS SDK provides its own thread pool). Please let me know if this satisfies your requirements.

eric-maynard commented on runtime/service/src/main/java/org/apache/polaris/service/events/jsonEventListener/JsonEventListener.java:
```diff
@@ -0,0 +1,36 @@
+import org.apache.polaris.service.events.AfterTableRefreshedEvent;
+import org.apache.polaris.service.events.PolarisEventListener;
+
+public abstract class JsonEventListener extends PolarisEventListener {
```
It might be nice to have a javadoc -- also this isn't really doing anything JSON-related

RussellSpitzer commented on runtime/service/src/main/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,181 @@
+    try {
+      CompletableFuture<CreateLogGroupResponse> future =
+          client.createLogGroup(CreateLogGroupRequest.builder().logGroupName(logGroup).build());
+      future.join();
```
Ah nvm, I see now that get isn't going to get the real exception anyway.

RussellSpitzer commented on runtime/service/src/main/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,181 @@
+    try {
+      eventAsJson = objectMapper.writeValueAsString(properties);
+    } catch (JsonProcessingException e) {
+      LOGGER.error("Error processing event into JSON string: {}", e.getMessage());
```
Why not use 

Logger.error(String str, Throwable err)?

adnanhemani replied:
Good call, added in next revision

RussellSpitzer commented on runtime/service/src/main/java/org/apache/polaris/service/quarkus/events/jsonEventListener/aws/cloudwatch/QuarkusAwsCloudWatchConfiguration.java:
```diff
@@ -0,0 +1,100 @@
+   *
+   * <p>Configuration property: {@code polaris.event-listener.aws-cloudwatch.log-group}
+   *
+   * @return an Optional containing the log group name, or empty if not configured
```
I may just not understand how qurakus does this, but why isn't it the default if not configured?

adnanhemani replied:
This is a good call - this may be on me due to hangover from a previous commit. Adjusting for the next revision.

RussellSpitzer replied:
Doc still needs update (same for below docs)

RussellSpitzer commented on runtime/service/src/test/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListenerTest.java:
```diff
@@ -0,0 +1,432 @@
+
+  enum TestMode {
+    LOCALSTACK,
+    MOCKED
```
For give me if this was from a previous review but I'm not sure I understand why "mocked" mode is useful here. It makes the test code very complicated and I'm not sure it tests very much. The LocalStack impl seems to be a much better test.

adnanhemani replied:
It was asked from a previous [review](https://github.com/apache/polaris/pull/1965#discussion_r2210737076). I think the main thing it does is give a way to test the code in a mocked fashion, which will be similar to other future CSP product implementations, as they do not have a LocalStack equivalence.

adnanhemani replied:
Talked offline with @eric-maynard - he is fine with removing the "mocked"-mode tests. Will reflect in the next revision.

RussellSpitzer commented on runtime/service/src/test/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListenerTest.java:
```diff
@@ -0,0 +1,310 @@
+            });
+
+    // Clean up
+    listener.shutdown();
```
Cleanups can go in an "afterEach" then you also won't have to handle exceptions and can just rethrow everything

RussellSpitzer replied:
I actually have other thoughts on this now, I think maybe you should do try finally? That way you would have to have class level references to the listener and client...

adnanhemani replied:
Sure, I've adjusted to use the try-finally for the tests.

adnanhemani commented on runtime/service/src/test/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListenerTest.java:
```diff
@@ -0,0 +1,310 @@
+      try {
+        Thread.sleep(100);
+      } catch (InterruptedException e) {
+        fail("Thread interrupted waiting for event");
```
Sure, I've adjusted to use the try-finally for the tests.

RussellSpitzer commented on runtime/service/src/test/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListenerTest.java:
```diff
@@ -0,0 +1,311 @@
+      GetLogEventsResponse logEvents;
+      int eventCount = 0;
+      long startTime = System.currentTimeMillis();
+      while (eventCount == 0) {
```
We can use [awaitility](https://github.com/awaitility/awaitility) here, it's already in the project so no big lift here

adnanhemani replied:
Good call - I've updated this in the next revision.

RussellSpitzer commented on runtime/service/src/test/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListenerTest.java:
```diff
@@ -0,0 +1,297 @@
+  }
+
+  @Test
+  void shouldCreateLogGroupAndStream() {
```
We are missing the negative test here, we also ned to test that it doesn't fail if the log and group already exists. You could just call the create a second time

RussellSpitzer reviewed (dismissed):
I think we are just missing one test of the "already exists" branch but other than that I think this is ready to go

eric-maynard commented on runtime/service/src/main/java/org/apache/polaris/service/events/jsonEventListener/JsonEventListener.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Abstract base class from which all event sinks that output events in JSON format can extend.
+ *
+ * <p>This class provides a common framework for transforming Polaris events into JSON format and
```
Does this actually transform events into a JSON format?

eric-maynard commented on runtime/service/src/main/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,158 @@
+
+  @Override
+  protected void transformAndSendEvent(HashMap<String, Object> properties) {
+    properties.put("realm", callContext.getRealmContext().getRealmIdentifier());
+    properties.put("principal", securityContext.getUserPrincipal().getName());
```
nit but I would recommend putting these into constants somewhere, e.g. `AwsCloudWatchEventEventListener.KEY_REALM`

singhpk234 commented on runtime/service/src/main/java/org/apache/polaris/service/events/jsonEventListener/aws/cloudwatch/AwsCloudWatchEventListener.java:
```diff
@@ -0,0 +1,158 @@
+  @Override
+  protected void transformAndSendEvent(HashMap<String, Object> properties) {
+    properties.put("realm", callContext.getRealmContext().getRealmIdentifier());
+    properties.put("principal", securityContext.getUserPrincipal().getName());
```
Agree, its not a must to have (not a blocker), we can add it later :), IMHO it would be better if we could have solidified the schema

singhpk234 reviewed (approved):
LGTM, Thanks @adnanhemani for all the work ! 

I have added some suggestions, which are optional to address and are good to have.

adnanhemani commented on site/content/in-dev/unreleased/configuration.md:
```diff
@@ -78,46 +78,51 @@ read-only mode, as Polaris only reads the configuration file once, at startup.
+| `polaris.config.rollback.compaction.on-conflicts.enabled`                              | `false`                            | When set to true Polaris will apply the deconfliction by rollbacking those REPLACE operations snapshots which have the property of `polaris.internal.rollback.compaction.on-conflict` in their snapshot summary set to `rollback`, to resolve conflicts at the server end.      ...
+| `polaris.event-listener.type`                                                          | `no-op`                            | Define the Polaris event listener type. Supported values are `no-op`, `aws-cloudwatch`.                                                                                                                                                                                         ...
+| `polaris.event-listener.aws-cloudwatch.log-group`                                      | `polaris-cloudwatch-default-group` | Define the AWS CloudWatch log group name for the event listener.                                                                                                                                                                                                                ...
+| `polaris.event-listener.aws-cloudwatch.log-stream`                                     | `polaris-cloudwatch-default-stream`| Define the AWS CloudWatch log stream name for the event listener. Ensure that Polaris' IAM credentials have the following actions: "PutLogEvents", "DescribeLogStreams", and "DescribeLogGroups" on the specified log stream/group. If the specified log stream/group does not  ...
```
I don't see one other than the Getting Started - and I don't think that page is relevant here.

adnanhemani commented on runtime/service/src/main/java/org/apache/polaris/service/events/jsonEventListener/PropertyMapEventListener.java:
```diff
@@ -0,0 +1,43 @@
+  public void onAfterTableRefreshed(AfterTableRefreshedEvent event) {
+    HashMap<String, Object> properties = new HashMap<>();
+    properties.put("event_type", event.getClass().getSimpleName());
+    properties.put("table_identifier", event.tableIdentifier().toString());
```
I agree with this but we should put this as part of the instrumentation of the event itself. I will look into this for #2480.

adnanhemani commented:
Working on rebase for this - seems like it is now failing in main
